conf/defaults.ini

```diff
@@ -770,6 +770,7 @@ allow_sign_up = true
 auto_login = false
 client_id = some_client_id
 client_secret =
+client_secret_jwt =
```

Comment:
Not sure why we would need this field. It looks like this one is used as a client id and we already have a field for that but maybe I am missing something?

We should have a configuration called `client_authentication` whose value can either be `client_secret_jwt` or `client_secret_post` and depending on that we should alter how the request to exchange authorization code to a token is made

Reply:
Hi Kalle,
When Web Apps are deployed in Microsoft Azure, on top of client secret/certificate, they may also be assigned a Managed Identity (since it is internally deployed). In order to authenticate, instead of using the client secret/cert, it uses the Client ID of the Managed Identity (different than client_id) to obtain a JWT token. My current implementation takes in this Managed Identity Client ID and obtains a JWT token which is what is used in the authentication flow.

I like your idea of creating a client_authentication config option that can hold a dynamic value of either client_secret_post or client_secret_jwt, but what if we have a predefined list of implemented options specific to each token provider. Say, for Microsoft Azure AD, we allow values of client_secret, and managed_identity (with their respective implementations)? Then, if the user defines client_authentication config option as managed_identity they then pass the Managed Identity Client ID as the client_secret value. Your thoughts/guidance is much appreciated!

Reply:
I see, while I think this is something we can support for the azure provider this is a azure specific feature. 

So the check here https://github.com/grafana/grafana/pull/95455/files#diff-856048eb43005b55a23bfa7c52b688572f712e0526a2d0c690f30e4fa1077387R139 is specific to that provider.

So I think we need to go about this a bit differently than what is currently done.

Instead of adding a new `GetClientSecretJWT` to `SocialConnector` interface we should perform this inside `Exchange`. Right now every Connector delegates to social base but we can add it to e.g. the azure one and perform these operations and then call the inner one. 

Ideally this would also support the standard e.g. using the secret provided to sign the token. But if this other setting is set we use the managed identity

Reply:
@kalleep Sounds good, I like that plan. Since, as you stated, this is a Microsoft Azure AD specific feature, should I rename the config from `client_secret_jwt` to something more Azure AD specific e.g. `msi_client_id`? Or should we have two flags, where one is `msi_auth_enabled` set to a boolean, and `msi_client_id` set to the Managed Identity Client ID (only used when `msi_auth_enabled` is set to true?

Reply:
Yes it think renaming that setting to something more descriptive would be good, in the azure data source we use [managed_identity_client_id](https://github.com/grafana/grafana/blob/003b2f14db0bec68a04a904c48558d34a51b7dbb/conf/defaults.ini#L964-L966) so lets go with that.

I have not looked into using managed identity yet but does it only work with `client_secret_jwt` or does it work with `client_secret_post` too?

I am assuming it only works with `client_secret_jwt` for now but if not wt can handle both cases. 

I have something like this in mind:
1. If `client_authentication = client_secret_post` we just authenticate as done right now
2. If `client_authentication = client_secret_jwt` we authenticate using the oidc spec e.g. sing the token with client_secret and include client_id in correct places, I think `iss` and `sub` must be set to client id.
2. If `lient_authentication = client_secret_jwt` _and_ `managed_identity_client_id != ""` we can sing the token using the managed identity as done currently

Reply:
Hi @kalleep,
Okay, `managed_identity_client_id` is good with me! 

Just to provide a brief overview again: The way it works is that we would pass the `managed_identity_client_id` in as a field, and then there is back end code [here](https://github.com/grafana/grafana/pull/95455/files#:~:text=func%20(s%20*SocialAzureAD)%20GetClientSecretJWT(ctx%20context.Context)%20(string%2C%20error)%20%7B) that retrieves the JWT token. This token is passed outright as the `client_assertion` [here](https://github.com/grafana/grafana/pull/95455/files#:~:text=opts%20%3D%20append(opts%2C%20oauth2.SetAuthURLParam(%22client_assertion%22%2C%20clientAssertion))).

Managed Identity is only supported by `client_secret_jwt` (which in fact is `client_assertion` when looking at the actual POST requests), so we would only need to support this case. Also, I am not sure about how to implement your (2) but I understand completely with your (1) and (3).

Under what section would we put the `client_authentication` flag in the .ini file?

Reply:
Yes I see that but we can do that all in the exchange method instead.

You would ideally just follow the standard spec: https://openid.net/specs/openid-connect-core-1_0.html#ClientAuthentication

I found some Microsoft specific documentation for this too:
https://learn.microsoft.com/en-us/entra/identity-platform/v2-oauth2-client-creds-grant-flow#second-case-access-token-request-with-a-certificate
https://learn.microsoft.com/en-us/entra/identity-platform/certificate-credentials

You would use the jwt package we use in grafana [go-jose](https://github.com/go-jose/go-jose) to manually build the token according to the spec and sign it with `client_secret` and the add the correct params to the url.

Reply:
@kalleep Sounds good! Where do you think we should place the `client_authentication` flag? I am assuming we would put the `managed_identity_client_id` flag under `[auth.azuread]`?

Reply:
Yes they should both be under `[auth.azuread]`.

What you linked me only seems to handle `private_key_jwt`, that is a different flow from `client_secret_jwt`.

You would essentially build the token as they do in the code that you linked but instead of signing it with a private key you would sing it using the `client_secret`.  At some point we could also support `private_key_jwt` but that is out of scope for the issue I created and that would be yet another option for `client_authentication`.

I would just suggest you follow the oidc spec.

Reply:
@kalleep Ah yes, I seem to have misunderstood the documentation. I believe I have all my questions answered and will update you when I have an updated PR for you to review!

Reply:
@kalleep Update:
So I just tested the implementation of the oidc spec with `client_secret_jwt` when `managed_identity_client_id` is null, and I got this error:
`ERROR[11-06|10:41:08] Failed to authenticate request           logger=authn.service client=auth.client.azuread error="[auth.oauth.token.exchange] failed to exchange code to token: oauth2: \"invalid_client\" \"AADSTS5002738: Invalid JWT token. 'HS256' is not a supported signature algorithm. Supported signing algorithms are: 'RS256,RS384,RS512,PS256,PS384,PS512' Trace ID: aa30e445-255c-49c3-b644-b10ddf269100 Correlation ID: ffa6b608-93ff-4af0-ace6-2b1161f512cf Timestamp: 2024-11-06 18:41:09Z\" \"https://login.microsoftonline.com/error?code=5002738\""`

It seems as though Microsoft no longer supports HS256 (symmetric key encryption), and only supports RS and PS (asymmetric key encryption), most likely due to security concerns. I do have all three implementations coded and will wait for your feedback on this error.